Let Process.spawn handle STDOUT redirection on MRI

afn opened the pull request:
Fixes #854

afn commented:
I couldn't come up with a great spec for this, but the one in the PR does demonstrate the bug. Running the spec without the bugfix causes it to (typically) fail with an error like:

```
  1) Capybara::Poltergeist::Driver is threadsafe in how it captures console.log
     Failure/Error: puts i

     IOError:
       stream closed
     # ./spec/integration/driver_spec.rb:62:in `write'
     # ./spec/integration/driver_spec.rb:62:in `puts'
     # ./spec/integration/driver_spec.rb:62:in `puts'
     # ./spec/integration/driver_spec.rb:62:in `block (4 levels) in <module:Poltergeist>'
     # ./spec/integration/driver_spec.rb:61:in `upto'
     # ./spec/integration/driver_spec.rb:61:in `block (3 levels) in <module:Poltergeist>'
```

afn commented:
The failing builds don't appear to be related to this change. (It looks like bundler is picking up a new version of ttfunk, released today, that depends on ruby ~> 2.1)

twalpole commented on lib/capybara/poltergeist/client.rb:
```diff
@@ -102,14 +103,20 @@ def command
-      STDOUT.reopen(prev)
-      $stdout = STDOUT
-      prev.close
+      if Capybara::Poltergeist.jruby?
```
Does rubinius support the :out option?

twalpole replied:
Also, I believe JRuby 9000 should support the :out option (could be wrong on that but I think so)

afn replied:
Good question. If not, we could make this test more strict.

Separately, the thread safety issue still exists for JRuby (and Rubinius, if it doesn't support the `:out` option). I think it's outside the scope of this PR, but it may be worth looking into an alternate solution there, e.g. `fork`, then redirect STDOUT, then `exec` (assuming those VMs support `fork` and `exec`) rather than `Process.spawn`.

route replied:
There is no fork for JVM

twalpole commented:
I fixed the ttfunk issue - please rebase so the tests can be fully run

route commented:
This code was exactly written to be working across different ruby implementations. I wouldn't change it without strong reason and if I do I'd do it carefully.

afn commented:
@route the current implementation definitely doesn't work correctly in multithreaded applications on MRI, so at a minimum, this should be fixed to use `:out` on MRI. I strongly suspect that the thread-safety issue exists on most, if not all, other platforms as well, so ideally we should change this patch to use `:out` everywhere that it's available (and nowhere that it's not).

On platforms that **don't** support `:out`, what happens if you pass it in? Is it silently ignored, or does `Process.spawn` raise an error? If it's the latter, I would argue that the ideal approach is to try passing `:out`, catch the error if it's thrown, and then fall back on the `redirect_stdout` approach ... after printing a warning message (one that can be disabled via configuration) informing users that it's not threadsafe.

route commented:
@afn how does your test fail without this patch?

afn commented:
@route our production app intermittently dies with `Errno::EPIPE: Broken pipe @ io_write - <STDOUT>`. I wasn't able to exactly reproduce *that* failure in a test, but the one I wrote fails pretty consistently with:

```
     IOError:
       stream closed
     # ./spec/integration/driver_spec.rb:62:in `write'
     # ./spec/integration/driver_spec.rb:62:in `puts'
     # ./spec/integration/driver_spec.rb:62:in `puts'
     # ./spec/integration/driver_spec.rb:62:in `block (4 levels) in <module:Poltergeist>'
     # ./spec/integration/driver_spec.rb:61:in `upto'
     # ./spec/integration/driver_spec.rb:61:in `block (3 levels) in <module:Poltergeist>'
```

twalpole commented on spec/integration/driver_spec.rb:
```diff
@@ -56,6 +56,29 @@ def session_url(path)
       end
     end
 
+    unless Capybara::Poltergeist.jruby?
```
instead of of not running the test when jruby, please set the test to pending when jruby is true so that we can see the results in the logs as an accepted failure

afn replied:
Hmm, looks like it doesn't actually fail in jruby. I'd be surprised if the thread-safety issue didn't exist there, though. Let me see if I can come up with a better test.

twalpole replied:
A test that doesn't actually add 100k lines to the log would be good too

afn replied:
😁

afn commented:
Updated the spec, so that it's deterministic; it's pending on jruby. I guess the only outstanding question is whether we can use `:out` on other rubies, e.g. Rubinius. What's the right condition for using `:out`?

route commented:
@afn @twalpole I haven't been following Rubinius for a while is it worth supporting now?

afn commented:
Looks like `:out` doesn't work on Rubinius either, so I've changed the code to use `redirect_stdout` on JRuby and Rubinius.

BTW, the build wasn't running at all on Rubinius, but the build failure was ignored. I've updated `.travis.yml` to get it to work. The build actually passes (with the new test pending), but it segfaults right after finishing running the tests, so unfortunately we still have to leave Rubinius in the `allow_failures` list.

twalpole commented:
@route The biggest issue with rubinius is getting it to run stably on travis (as @afn found out)

twalpole commented on spec/integration/driver_spec.rb:
```diff
@@ -56,6 +56,54 @@ def session_url(path)
+
+    it 'is threadsafe in how it captures console.log' do
+      stdout = capture_stdout do
+        pending if Capybara::Poltergeist.jruby? || Capybara::Poltergeist.rubinius?
```
Please add a descriptive text here -  `pending "blah blah not supported" if ...`

twalpole commented on lib/capybara/poltergeist/client.rb:
```diff
@@ -64,6 +64,7 @@ def start
 
       process_options = {}
       process_options[:pgroup] = true unless Capybara::Poltergeist.windows?
+      process_options[:out] = @write_io unless Capybara::Poltergeist.jruby?
```
Should this check rubinius? too?

afn replied:
Indeed

twalpole commented on spec/integration/driver_spec.rb:
```diff
@@ -56,6 +56,54 @@ def session_url(path)
       end
     end
 
+    def capture_stdout
```
I think this can be replaced with RSpec's output matcher - https://www.relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/output-matcher

afn replied:
Nice! I didn't know about that.

twalpole commented:
JRuby 9.1.7.0 appears to pass the correctly if the `:out` option is passed - so either it supports :out or the test isn't valid

afn commented:
Is there a way to test whether the current JRuby version is new enough to
support `:out`?
On Mon, Feb 20, 2017 at 4:01 PM Thomas Walpole <notifications@github.com>
wrote:

> JRuby 9.1.7.0 appears to pass the correctly if the :out option is passed
> - so either it supports :out or the test isn't valid
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/teampoltergeist/poltergeist/pull/855#issuecomment-281180429>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACppHlzGg_wCspIEu_X-HR2DamUJbky6ks5ref82gaJpZM4L_wSW>
> .
>

twalpole commented:
There's the global constant `JRUBY_VERSION` but I haven't been able to find any actual documentation on what version should work correctly with :out and what shouldn't. I'm also curious whether it's platform specific behavior (linux/macos/windows) if anyone knows?

afn commented:
Well how about we merge this PR, which is a conservative fix (in that it fixes the issue on MRI, but not on JRuby or Rubinius; but it also doesn't introduce regressions on those platforms), and then open a separate issue to address the thread-unsafe behavior on JRuby and Rubinius. Then as part of triaging *that* issue, we can look into whether the bug is OS-specific, whether Rubinius is still of interest, which version of JRuby added support for `:out`, etc. Does that sound OK?

twalpole commented:
My concern here would also be whether it has platform specific issues on MRI - @route?

route commented:
@twalpole I can't recall MRI's issues, but I definitely remember how it failed on jruby or rubinius either. Moreover I tried to fix it different ways including out option and when it worked on jruby then rubinius failed and vice versa haha) But time has passed since then and maybe languages has changed decently. Here's the related PR https://github.com/teampoltergeist/poltergeist/pull/405

twalpole commented:
@afn Could you please change this to check for MRI to opt in to the new behavior rather than checking JRuby and rubinius to opt out of it and then squash the commits.

afn commented:
@twalpole Done - LMK how that looks. Thanks!

twalpole commented:
Thanks!
